packages/code-infra/src/withDeploymentConfig.mjs

```diff
@@ -0,0 +1,86 @@
+/**
+ * See the docs of the Netlify environment variables:
+ * https://docs.netlify.com/configure-builds/environment-variables/#build-metadata.
+ *
+ * A few comments:
+ * - process.env.CONTEXT === 'production' means that the branch in Netlify was configured as production.
+ *   For example, the `master` branch of the Core team is considered a `production` build on Netlify based
+ *   on https://app.netlify.com/sites/material-ui/settings/deploys#branches.
+ * - Each team has different site https://app.netlify.com/teams/mui/sites.
+ *   The following logic must be compatible with all of them.
+ */
+let DEPLOY_ENV = 'development';
+
+// Same as process.env.PULL_REQUEST_ID
+if (process.env.CONTEXT === 'deploy-preview') {
+  DEPLOY_ENV = 'pull-request';
+}
+
+if (process.env.CONTEXT === 'production' || process.env.CONTEXT === 'branch-deploy') {
+  DEPLOY_ENV = 'production';
+}
+
+// The 'master' and 'next' branches are NEVER a production environment. We use these branches for staging.
+if (
+  (process.env.CONTEXT === 'production' || process.env.CONTEXT === 'branch-deploy') &&
+  (process.env.HEAD === 'master' || process.env.HEAD === 'next')
+) {
+  DEPLOY_ENV = 'staging';
+}
+/**
+ * ====================================================================================
+ */
+
+process.env.DEPLOY_ENV = DEPLOY_ENV;
+
+/**
+ * @typedef {import('next').NextConfig} NextConfig
+ */
+
+/**
+ *
+ * @param {NextConfig} nextConfig
+ * @returns {NextConfig}
+ */
+export function withDeploymentConfig(nextConfig) {
```

Comment:
I like the idea of extracting the deployment config as a specific next.js plugin. Makes it nice and contained.